File: src/main/java/fr/yodamad/svn2git/service/util/Shell.java
package fr.yodamad.svn2git.service.util;

import fr.yodamad.svn2git.domain.Migration;
import org.apache.commons.io.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.nio.charset.Charset;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.Executors;
import java.util.function.Consumer;

import static java.lang.String.format;

/**
 * Shell utilities
 */
public abstract class Shell {

    private static final Logger LOG = LoggerFactory.getLogger(Shell.class);

    public static final boolean isWindows = System.getProperty("os.name").toLowerCase().startsWith("windows");

    /**
     * Get working directory
     * @param directory
     * @param mig
     * @return
     * @throws InterruptedException
     * @throws IOException
     */
    public static String workingDir(String directory, Migration mig) throws IOException, InterruptedException, RuntimeException {
        String today = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
        if (isWindows) {
            String path = format("%s%s", System.getenv("SystemDrive"), directory.replaceAll("/", "\\\\"));
            if (!new File(path).exists()) {
                String mkdir = format("mkdir %s", path);
                execCommand(System.getenv("SystemDrive"), mkdir);
            }
            return directory + "\\" + today + "_" + mig.getId();
        }
        if (!new File(directory).exists()) {
            String mkdir = format("mkdir -p %s", directory);
            execCommand("/", mkdir);
        }
        return directory + "/" + today + "_" + mig.getId();
    }

    /**
     * Get git working directory
     * @param root
     * @param sub
     * @return
     */
    public static String gitWorkingDir(String root, String sub) {
        if (isWindows) {
            return Shell.formatDirectory(root + "\\" + sub);
        }
        return root + "/" + sub;
    }

    /**
     * Execute a commmand through process
     * @param directory Directory in which running command
     * @param command command to execute
     * @throws InterruptedException
     * @throws IOException
     */
    public static int execCommand(String directory, String command) throws InterruptedException, IOException {
        return execCommand(directory, command, command);
    }

    /**
     * Execute a commmand through process without an alternative command to print in history
     * @param directory Directory in which running command
     * @param command command to execute
     * @param securedCommandToPrint command to print in history without password/token/...
     * @throws InterruptedException
     * @throws IOException
     */
    public static int execCommand(String directory, String command, String securedCommandToPrint) throws InterruptedException, IOException {
        ProcessBuilder builder = new ProcessBuilder();
        String execDir = formatDirectory(directory);
        if (isWindows) {
            builder.command("cmd.exe", "/c", command);
        } else {
            builder.command("sh", "-c", command);
        }

        builder.directory(new File(execDir));

        LOG.debug(format("Exec command : %s", securedCommandToPrint));
        LOG.debug(format("in %s", execDir));

        Process process = builder.start();
        StreamGobbler streamGobbler = new StreamGobbler(process.getInputStream(), LOG::debug);
        Executors.newSingleThreadExecutor().submit(streamGobbler);

        StreamGobbler errorStreamGobbler = new StreamGobbler(process.getErrorStream(), LOG::debug);
        Executors.newSingleThreadExecutor().submit(errorStreamGobbler);

        String stderr = IOUtils.toString(process.getErrorStream(), Charset.defaultCharset());

        int exitCode = process.waitFor();
        LOG.debug(format("Exit : %d", exitCode));

        if (exitCode != 0) {
            throw new RuntimeException(stderr);
        }

        return exitCode;
    }

    // Utils
    private static class StreamGobbler implements Runnable {
        private InputStream inputStream;
        private Consumer<String> consumer;

        StreamGobbler(InputStream inputStream, Consumer<String> consumer) {
            this.inputStream = inputStream;
            this.consumer = consumer;
        }

        @Override
        public void run() {
            new BufferedReader(new InputStreamReader(inputStream)).lines()
                .forEach(consumer);
        }
    }

    /**
     * Format directory to fit f*** windows behavior
     * @param directory Directory to format
     * @return formatted directory
     */
    public static String formatDirectory(String directory) {
        String execDir = directory;
        if (isWindows) {
            execDir = directory.startsWith("/") ?
                format("%s%s", System.getenv("SystemDrive"), directory).replaceAll("/", "\\\\")
                : directory;
        }
        return execDir;
    }
}

File: src/main/kotlin/fr/yodamad/svn2git/io/Shell.kt
package fr.yodamad.svn2git.io

import fr.yodamad.svn2git.domain.Migration
import fr.yodamad.svn2git.service.util.CommandManager
import org.apache.commons.io.IOUtils
import org.slf4j.LoggerFactory
import java.io.*
import java.nio.charset.Charset
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.Executors
import java.util.function.Consumer

/**
 * Shell utilities
 */
object Shell {
    private val LOG = LoggerFactory.getLogger(Shell::class.java)
    val isWindows = System.getProperty("os.name").toLowerCase().startsWith("windows")

    /**
     * Get working directory
     * @param directory
     * @param mig
     * @return
     * @throws InterruptedException
     * @throws IOException
     */
    @Throws(IOException::class, InterruptedException::class, RuntimeException::class)
    fun workingDir(commandManager: CommandManager, directory: String, mig: Migration): String {
        val today = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
        if (isWindows) {
            val path = String.format("%s%s", System.getenv("SystemDrive"), directory.replace("/".toRegex(), "\\\\"))
            if (!File(path).exists()) {
                val mkdir = String.format("mkdir %s", path)
                execCommand(commandManager, System.getenv("SystemDrive"), mkdir)
            }
            return directory + "\\" + today + "_" + mig.id
        }
        if (!File(directory).exists()) {
            val mkdir = String.format("mkdir -p %s", directory)
            execCommand(commandManager, "/", mkdir)
        }
        return directory + "/" + today + "_" + mig.id
    }

    /**
     * Get git working directory
     * @param root
     * @param sub
     * @return
     */
    fun gitWorkingDir(root: String, sub: String): String {
        return if (isWindows) {
            formatDirectory(root + "\\" + sub)
        } else "$root/$sub"
    }

    /**
     * Execute a commmand through process
     * @param directory Directory in which running command
     * @param command command to execute
     * @throws InterruptedException
     * @throws IOException
     */
    @JvmOverloads
    @Throws(InterruptedException::class, IOException::class)
    fun execCommand(commandManager: CommandManager, directory: String, command: String?, securedCommandToPrint: String? = command): Int {
        val builder = ProcessBuilder()
        val execDir = formatDirectory(directory)
        if (isWindows) {
            builder.command("cmd.exe", "/c", command)
        } else {
            builder.command("sh", "-c", command)
        }
        builder.directory(File(execDir))
        LOG.debug(String.format("Exec command : %s", securedCommandToPrint))
        LOG.debug(String.format("in %s", execDir))
        val process = builder.start()
        val streamGobbler = StreamGobbler(process.inputStream) { s: String? -> LOG.debug(s) }
        Executors.newSingleThreadExecutor().submit(streamGobbler)
        val errorStreamGobbler = StreamGobbler(process.errorStream) { s: String? -> LOG.debug(s) }
        Executors.newSingleThreadExecutor().submit(errorStreamGobbler)
        val stderr = IOUtils.toString(process.errorStream, Charset.defaultCharset())
        val exitCode = process.waitFor()
        LOG.debug(String.format("Exit : %d", exitCode))
        if (exitCode != 0) {
            // trace failed commands
            if (securedCommandToPrint != null) {
                commandManager.addFailedCommand(directory, securedCommandToPrint, stderr)
            }
            throw RuntimeException(stderr)
        } else {
            // trace successful commands
            if (securedCommandToPrint != null) {
                commandManager.addSuccessfulCommand(directory, securedCommandToPrint)
            }
        }
        return exitCode
    }

    /**
     * Format directory to fit f*** windows behavior
     * @param directory Directory to format
     * @return formatted directory
     */
    fun formatDirectory(directory: String): String {
        var execDir = directory
        if (isWindows) {
            execDir = if (directory.startsWith("/")) String.format("%s%s", System.getenv("SystemDrive"), directory).replace("/".toRegex(), "\\\\") else directory
        }
        return execDir
    }

    // Utils
    private class StreamGobbler constructor(private val inputStream: InputStream, private val consumer: Consumer<String?>) : Runnable {
        override fun run() {
            BufferedReader(InputStreamReader(inputStream)).lines()
                .forEach(consumer)
        }
    }
}
